qiime2/core/util.py

```
@@ -470,3 +471,107 @@ def create_collection_name(*, name, key, idx, size):
         standardized way. Assumes 0 based indexing.
     """
     return [name, key, f'{idx + 1}/{size}']
+
+
+def replace_bytes_in_directory(directory, old_bytes, new_bytes, extensions,
+                               buffer_size=None):
+    """
+    Recursively replaces all occurrences of old_bytes with new_bytes in files
+    under the given directory. Only processes files with specified extensions.
+
+    Parameters
+    ----------
+    directory : str
+        The root directory to recursively search.
+    old_bytes : bytes
+        The byte sequence to find and replace. Must be at least 2 bytes long.
+    new_bytes : bytes
+        The byte sequence to replace old_bytes with.
+    extensions : set[str] or list[str]
+        File extensions to process (e.g., {'.txt', '.md'}).
+        Extensions should include the leading dot.
+    buffer_size : int, optional
+        Size of the read buffer in bytes. Default is 4096.
+    """
+    extensions = set(extensions)
+
+    for root, _, files in os.walk(directory):
+        for file in files:
+            if os.path.splitext(file)[1] in extensions:
+                filepath = os.path.join(root, file)
+                replace_bytes_in_file(filepath, old_bytes, new_bytes,
+                                      buffer_size)
+
+
+def replace_bytes_in_file(filepath, old_bytes, new_bytes, buffer_size=None):
+    """
+    Replaces all occurrences of old_bytes with new_bytes in the specified file.
+
+    Parameters
+    ----------
+    filepath : str
+        Path to the file to process.
+    old_bytes : bytes
+        The byte sequence to find and replace. Must be at least 2 bytes long.
```

Why does this have to be at least two bytes? And if this is a hard requirement, should we raise an error if it isn't met?

---

I think it had more to do with this being a silly function to use if you only had a string of 1 or less to replace. None of the regex business is needed anymore in that case.

Also it's a `qiime2.core.util` so I didn't worry about the sharper surface of the API.

Let me know if you would want me to change this.

---

yeah it can handle 1 byte also, but this function is just an unreasonably expensive way to do it

---

That's about what I figured. I did take some time thinking about what mechanism might actually make it not work if it was 1 byte, and I couldn't come up with one because there isn't one. There should probably be a disclaimer saying, "This will work with 1 byte, it's just a silly way of doing it with 1 byte" if only to save anyone in the future from sitting down and trying to come up with a technical reason it would break with 1 byte.